### PR TITLE
Override `schema_format` per database via configuration

### PR DESCRIPTION
**Pull request by leboshi**
### Summary

Partially fixes (or rather implements) #43173.  (I'd be happy to tackle the other half of that feature request, depending on the feedback on this half.)

Each database can specify its own `schema_format` for dumping/loading in the database configuration (e.g., `database.yml`).  `ActiveRecord.schema_format` can still be set as a global default and is used for any database configurations that don't specify their own `schema_format`.

### Breaking Changes

Because the schema format is now bundled into the database configuration, passing the format as a separate argument doesn't make any sense.  I've removed that argument from methods that accept a database configuration.  ActiveRecord itself is fine, but anyone calling these methods from external code will need to update it.  If we should move more slowly, I can add a deprecation warning and do some fun argument detection and shuffling, a la `link_to`...  Not sure where the balance is between the complexity of doing that and the friendliness of shimming the transition.

Because the default schema format has been moved from a default argument value into the database configuration, it also made sense to me to deprecate omitting that argument.  Depending on where this feature goes on the Rails roadmap, the specified horizon in that deprecation warning can be changed.

### CODE REVIEW

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,15 @@
...
+    Allow each database configuration to specify a `schema_format`, and move the default value of
+    `ActiveRecord.schema_format` to the configuration accessor from default parameter values.  This allows removal of 
+    the `format` parameter from methods that already take in a database configuration, but also makes it necessary to
+    pass the format from the configuration (which may be the default) to methods where it was previously optional.
+
+    Fixes #43173
+
+    *Ryan Kerr*
+
+* Add support for generated columns in PostgreSQL adapter
```
Comment (rafaelfranca) [inline activerecord/CHANGELOG.md:12]:
```suggestion
*   Add support for generated columns in PostgreSQL adapter
```

File: activerecord/CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,15 @@
-*   Add support for generated columns in PostgreSQL adapter
+* Override `schema_format` per database via configuration
```
Comment (rafaelfranca) [inline activerecord/CHANGELOG.md:1]:
```suggestion
*   Override `schema_format` per database via configuration
```

File: activerecord/CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,15 @@
...
+
+    Allow each database configuration to specify a `schema_format`, and move the default value of
+    `ActiveRecord.schema_format` to the configuration accessor from default parameter values.  This allows removal of
```
Comment (rafaelfranca) [inline activerecord/CHANGELOG.md:4]:
```suggestion
    `ActiveRecord.schema_format` to the configuration accessor from default parameter values. This allows removal of 
```

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 4]
```diff
@@ -399,7 +399,7 @@ def load_schema(db_config, format = ActiveRecord.schema_format, file = nil) # :n
         Migration.verbose = verbose_was
       end
 
-      def schema_up_to_date?(configuration, format = ActiveRecord.schema_format, file = nil, environment = nil, name = nil)
+      def schema_up_to_date?(configuration, file = nil, environment = nil, name = nil)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/tasks/database_tasks.rb:402]:
```suggestion
      def schema_up_to_date?(configuration, file = nil, environment = nil, name = nil) # :nodoc:
```

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 5]
```diff
@@ -460,12 +461,30 @@ def dump_schema(db_config, format = ActiveRecord.schema_format) # :nodoc:
         end
       end
 
-      def schema_file(format = ActiveRecord.schema_format)
+      def schema_file(format = nil)
+        unless format
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Omitting a schema format is deprecated and will raise an error in Rails 7.0.
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/tasks/database_tasks.rb:467]:
```suggestion
            Omitting a schema format is deprecated and will raise an error in Rails 7.1.
```

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 6]
```diff
@@ -460,12 +461,30 @@ def dump_schema(db_config, format = ActiveRecord.schema_format) # :nodoc:
...
+
         File.join(db_dir, schema_file_type(format))
       end
       deprecate :schema_file
 
-      def schema_file_type(format = ActiveRecord.schema_format)
+      def schema_file_type(format = nil)
+        unless format
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Omitting a schema format is deprecated and will raise an error in Rails 7.0.
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/tasks/database_tasks.rb:481]:
```suggestion
            Omitting a schema format is deprecated and will raise an error in Rails 7.1.
```

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 7]
```diff
@@ -474,7 +493,16 @@ def schema_file_type(format = ActiveRecord.schema_format)
         end
       end
 
-      def dump_filename(db_config_name, format = ActiveRecord.schema_format)
+      def dump_filename(db_config_name, format = nil)
+        unless format
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Omitting a schema format is deprecated and will raise an error in Rails 7.0.
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/tasks/database_tasks.rb:499]:
```suggestion
            Omitting a schema format is deprecated and will raise an error in Rails 7.1.
```

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 8]
```diff
@@ -494,9 +522,9 @@ def cache_dump_filename(db_config_name, schema_cache_path: nil)
         schema_cache_path || ENV["SCHEMA_CACHE"] || File.join(ActiveRecord::Tasks::DatabaseTasks.db_dir, filename)
       end
 
-      def load_schema_current(format = ActiveRecord.schema_format, file = nil, environment = env)
+      def load_schema_current(file = nil, environment = env)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/tasks/database_tasks.rb:525]: We need to deprecate the `format` options that was remove here.

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/tasks/database_tasks.rb (hunk 8 above)
Comment (leboshi) [reply activerecord/lib/active_record/tasks/database_tasks.rb:525]: Thank you so much for your feedback!  I apologize for the formatting errors.

I'm about to commit one idea for the deprecation here.  Let me know if you'd like me to do it differently.  Thanks again!

**Comment by leboshi** [comment]
Hey @eileencodes!  Sorry, I don't know how I missed all this.  Thank you so much for sharing the credit here.  That's extremely considerate and generous of you!  FWIW, I think your solution is a lot more elegant and comprehensive anyway.  Awesome work!

**Pull request closed without merging**